samples/katib/katib.py

```diff
@@ -1,7 +1,138 @@
+                                "python3",
+                                "/opt/mxnet-mnist/mnist.py",
+                                "--batch-size=64",
+                                "--lr=${trialParameters.learningRate}",
+                                "--num-layers=${trialParameters.numberLayers}",
+                                "--optimizer=${trialParameters.optimizer}"
+                            ]
+                        }
+                    ],
+                    "restartPolicy": "Never"
+                }
+            }
+        }
+    }
+
+    # Configure parameters for the Trial template.
+    # We can set the retain parameter to "True" to not clean-up the Trial Job's Kubernetes Pods.
+    trial_template = V1beta1TrialTemplate(
+        # retain=True,
+        primary_container_name="training-container",
+        success_condition='status.conditions.#(type=="Succeeded")#|#(status=="True")#',
+        failure_condition='status.conditions.#(type=="Failed")#|#(status=="True")#',
+        trial_parameters=[
+            V1beta1TrialParameterSpec(
+                name="learningRate",
+                description="Learning rate for the training model",
+                reference="lr"
+            ),
+            V1beta1TrialParameterSpec(
+                name="numberLayers",
+                description="Number of training model layers",
+                reference="num-layers"
+            ),
+            V1beta1TrialParameterSpec(
+                name="optimizer",
+                description="Training model optimizer (sdg, adam or ftrl)",
+                reference="optimizer"
+            ),
+        ],
+        trial_spec=trial_spec
+    )
+    return trial_template
+
+
+def experiment_search_params():
+    # Experiment search space.
+    # In this example we tune learning rate, number of layers and optimizer(sgd/adam/ftrl).
+    parameters = [
+        V1beta1ParameterSpec(
+            name="lr",
+            parameter_type="double",
+            feasible_space=V1beta1FeasibleSpace(
+                min="0.01",
+                max="0.03"
+            ),
+        ),
+        V1beta1ParameterSpec(
+            name="num-layers",
+            parameter_type="int",
+            feasible_space=V1beta1FeasibleSpace(
+                min="2",
+                max="5"
+            ),
+        ),
+        V1beta1ParameterSpec(
+            name="optimizer",
+            parameter_type="categorical",
+            feasible_space=V1beta1FeasibleSpace(
+                list=['sgd', 'adam', 'ftrl']
+            )
+        )
+    ]
+
+
+def algorithm_spec():
+    # Algorithm specification.
+    V1beta1AlgorithmSpec(
+        algorithm_name="random"
+    )
+
+def objective_spec(goal):
+    # Objective specification.
+    objective = V1beta1ObjectiveSpec(
+        type="maximize",
+        goal=goal,
+        objective_metric_name="Validation-accuracy",
+        additional_metric_names=["accuracy"]
+    )
```

Comment:
I think you should return the objective if you want to use function to create `objective_spec`, `algorithm_spec`, `experiment_search_params` and `generate_trial_template`.

```suggestion
def objective_spec(goal):
    # Objective specification.
    objective = V1beta1ObjectiveSpec(
        type="maximize",
        goal=goal,
        objective_metric_name="Validation-accuracy",
        additional_metric_names=["accuracy"]
    )
    return objective
```

Because of that, your Experiment has an empty objective and you have the nil pointer error that you mentioned here:
https://github.com/kubeflow/katib/issues/1435

Reply:
@andreyvelich Thank you for taking a look, your observation is correct and after fixing these the error does go away. This is one of the 🤦 moment for me, I am very sorry for making the silly mistake. However, is there a way to add a check to katib, so that a proper error is generated instead of a server panic.

Reply:
No worries @ScrapCodes.
Sure, I will add the check.